File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/core/Identifier.java
package app.zxtune.core;

import android.net.Uri;
import android.text.TextUtils;

import androidx.annotation.Nullable;

public class Identifier {

  private static final String SUBPATH_DELIMITER = "/";
  public static final Identifier EMPTY = new Identifier(Uri.EMPTY);

  /**
   * Identifier is not fully compatible with playlists from desktop version of zxtune
   * <p>
   * E.g. first track in ay file is stored as test.ay?#1 without hash sign encoding
   */
  private final Uri fullPath;
  private final Uri location;
  private final String subPath;

  public Identifier(Uri location, @Nullable String subPath) {
    final Uri.Builder builder = location.buildUpon();
    builder.fragment(null);
    this.location = builder.build();
    if (TextUtils.isEmpty(this.location.getPath())) {
      this.fullPath = this.location;
      this.subPath = "";
    } else {
      builder.fragment(subPath);
      this.fullPath = builder.build();
      String fragment = fullPath.getFragment();
      this.subPath = fragment != null ? fragment : "";
    }
  }

  public Identifier(Uri fullPath) {
    this(fullPath, fullPath.getFragment());
  }

  public static Identifier parse(@Nullable String str) {
    return str != null
            ? new Identifier(Uri.parse(str))
            : EMPTY;
  }

  public final Uri getFullLocation() {
    return fullPath;
  }

  public final Uri getDataLocation() {
    return location;
  }

  public final String getSubPath() {
    return subPath;
  }

  public final String getDisplayFilename() {
    final String filename = location.getLastPathSegment();
    if (subPath.length() == 0) {
      return filename != null ? filename : "";
    } else {
      final String subName = subPath.substring(subPath.lastIndexOf(SUBPATH_DELIMITER) + 1);
      return filename + " > " + subName;
    }
  }

  @Override
  public String toString() {
    return fullPath.toString();
  }

  @Override
  public int hashCode() {
    return fullPath.hashCode();
  }

  @Override
  public boolean equals(Object rh) {
    if (rh instanceof Identifier) {
      final Identifier id = (Identifier) rh;
      return fullPath.equals(id.fullPath);
    } else {
      return false;
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/core/Identifier.kt
package app.zxtune.core

import android.net.Uri

/**
 * Identifier is not fully compatible with playlists from desktop version of zxtune
 * <p>
 * E.g. first track in ay file is stored as test.ay?#1 without hash sign encoding
 */
class Identifier private constructor(builder: Uri.Builder) {

    val fullLocation: Uri = builder.build()

    val dataLocation: Uri = builder.fragment(null).build()

    val subPath: String
        get() = fullLocation.fragment.orEmpty()

    val displayFilename: String
        get() {
            val filename = dataLocation.lastPathSegment.orEmpty()
            return subPath.extractFilename { !it.isPackedDataIdentifier }?.let { nestedName ->
                "$filename > $nestedName"
            } ?: filename
        }

    val virtualFilename
        get() = subPath.extractFilename { !it.isPackedDataIdentifier && !it.isTrackIndexIdentifier }
            ?: dataLocation.lastPathSegment.orEmpty()

    val trackIndex: Int?
        get() = subPath.substringAfterLast("${SUBPATH_DELIMITER}${TRACK_INDEX_PREFIX}", "")
            .toIntOrNull()

    constructor(location: Uri) : this(
        if (location.path.isNullOrEmpty()) {
            Uri.EMPTY.buildUpon()
        } else {
            location.buildUpon()
        }
    )

    constructor(location: Uri, subPath: String?) : this(
        if (location.path.isNullOrEmpty()) {
            Uri.EMPTY.buildUpon()
        } else {
            location.buildUpon().fragment(subPath)
        }
    )

    override fun toString() = fullLocation.toString()

    override fun hashCode() = fullLocation.hashCode()

    override fun equals(other: Any?) = (other as? Identifier)?.fullLocation == fullLocation

    companion object {
        const val SUBPATH_DELIMITER = '/'
        const val TRACK_INDEX_PREFIX = '#'

        @JvmField
        val EMPTY = Identifier(Uri.EMPTY)

        @JvmStatic
        fun parse(str: String?) = str?.let { Identifier(Uri.parse(it)) } ?: EMPTY
    }
}

private val CharSequence.isPackedDataIdentifier
    get() = startsWith("+un")

private val CharSequence.isTrackIndexIdentifier
    get() = startsWith(Identifier.TRACK_INDEX_PREFIX)

private fun CharSequence.splitAsPath() =
    when (val lastDelimiter = lastIndexOf(Identifier.SUBPATH_DELIMITER)) {
        -1 -> "" to this
        else -> subSequence(0, lastDelimiter) to subSequence(lastDelimiter + 1, length)
    }

private fun CharSequence.extractFilename(filter: (CharSequence) -> Boolean): CharSequence? =
    splitAsPath().run {
        when {
            second.isEmpty() -> null
            filter(second) -> second
            first.isEmpty() -> null
            else -> first.extractFilename(filter)
        }
    }
